### ISSUE TITLE
Question on speed for ResNet 50 on single-GPU

### ISSUE THREAD

**Issue by Remember2018** [tensorpack/benchmarks#10]
Hi Yuxin, in your benchmarks, the speed of Tensorpack is 333 imgs per second. 

ResNet50 on fake ImageNet | 333 | 266

My question is, is this the limit of ResNet50 on single-GPU? Thanks.

**Comment by ppwwyyxx** [comment]
With tensorflow's XLA or other optimizations not available in tensorflow it should be able to train faster.

**Comment by Remember2018** [comment]
Hi Yuxin, @ppwwyyxx thanks for your comment! In fact, I'm also training a 2-class image classification using ResNet50 based on `ResNet` module in Tensorpack. There are about 600000 images with different sizes. I've tried two types of dataflow in your ResNet example. The first is directly reading images, and preprocess them and batch them. When I set batch_size as 256 uisng single GPU (V100, 16G), it seems to have 1.4iters/s, i.e., about 350 imgs/s. But the GPU utilization seems slow with much time of 0% usage. 

Becuase I'm afraid of too heavy concurrent reading operations on hard disk, so I'm resorting to the database solution. In your excellent dataflow, which provides excellent data prefetching and data loading, I'm not sure if this will have too heavy loads on hard disk. So I'm trying the second method in your ResNet example, i.e., using TFRecord format. First create the tfrecord files and based on the `tf.data.Dataset`, preprocess them and batch them. In this way, the speed seems still 1.4iters/s. And I checked this benchmark, which shows that in ResNet50 with Fake-ImageNet data, the speed is also 333 imgs/s. So I have this question. It seems the preprocessing (online data augmentation) costs too much time. But it is expected that the GPU utilization should be full. 

I'm also trainning the EfficientNet, it seems the speed is also about 300 imgs/s in single GPU(V100, 16G). 

I didn't use the XLA, which seems a compiler module in tensorflow. Does it have performance improvement on executation? I'll have a try if I had more GPUs. Thank you Yuxin.

**Comment by ppwwyyxx** [comment]
Since you're using different batch size (and perhaps different image size as well) you cannot compare the training speed with what's in this repo.

**Comment by Remember2018** [comment]
@ppwwyyxx yes, the input size are variable but the input of the network (after some data augmentations) are 224x224. I've found the bottleneck is in the data augmentation. I've pickled all the data and all original images are loaded in the memory. It is expected that the speed is faster. But no. So the bottleneck is in the data augmentation, not the disk reading. After I removed some data augmentation operations, Tensorpack can achieve about 4.5*256=1000+ imgs/s. I think offline augmentation would be better.
